Page freezes with many DOM manipulations and complex :has() selectors present

nipunshukla21 opened the pull request:
#### 4ef19b5fa4c093b11b221c8d8df3b5a3d9dd398d
<pre>
Page freezes with many DOM manipulations and complex :has() selectors present
<a href="https://bugs.webkit.org/show_bug.cgi?id=281920">https://bugs.webkit.org/show_bug.cgi?id=281920</a>
<a href="https://rdar.apple.com/138431700">rdar://138431700</a>

Reviewed by Matthieu Dubet and Antti Koivisto.

When evaluating :has(&gt; :is(.x &gt; .y)), child combinators within :is() are
scoped to the direct child&apos;s tree and don&apos;t require scope-breaking invalidation.
This avoids excessive DOM traversal on pages with many complex :has() selectors,
preventing page freezes during style recalculation.

* LayoutTests/fast/selectors/has-scope-breaking-classification-expected.txt:
* LayoutTests/fast/selectors/has-scope-breaking-classification.html:
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/invalidation/has-with-is-child-combinator-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/invalidation/has-with-is-child-combinator.html: Added.
* PerformanceTests/CSS/HasWithChildCombinatorInIs.html: Added.
* Source/WebCore/style/RuleFeature.cpp:
(WebCore::Style::computeNextHasPseudoClassMatchElement):

Canonical link: <a href="https://commits.webkit.org/302587@main">https://commits.webkit.org/302587@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c42161cde4772e20bb656aff49fa26e66d95d3e2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129312 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1568 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40149 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136689 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80704 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131183 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1780 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1445 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98482 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66383 "Found 1 new test failure: webgl/2.0.y/conformance2/query/occlusion-query.html (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132259 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1382 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115823 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79124 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1298 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33944 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79966 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109774 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34445 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139161 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1357 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1317 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107009 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1401 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112159 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106843 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27264 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1109 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30680 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/53991 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1429 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64793 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1247 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1282 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1351 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/style/RuleFeature.cpp:
```diff
@@ -195,8 +195,15 @@ static MatchElement computeNextHasPseudoClassMatchElement(MatchElement matchElem
     // `:has(:is(foo bar))` can be affected by changes outside the :has scope.
-    if (relation == CSSSelector::Relation::DescendantSpace || relation == CSSSelector::Relation::Child)
+    if (relation == CSSSelector::Relation::DescendantSpace || relation == CSSSelector::Relation::Child) {
+        // However, for `:has(> :is(.x > .y))`, the child combinator (>) inside :is() is still scoped to the direct child's tree.
+        // The parent in the relationship must be the direct child itself, which is within the :has(>) scope.
+        // Only descendant combinators can reach outside this scope (to ancestors of the subject element).
+        if (matchElement == MatchElement::HasChild && relation == CSSSelector::Relation::Child)
+            return matchElement;
```
nice!

mdubet reviewed (approved):
The change is correct and should help a lot 👍 

The test file should not use @scope though, it could be based on this test for example https://github.com/web-platform-tests/wpt/blob/4f26de067e0a83751bf0102f3a944ebb492b095b/css/selectors/invalidation/has-sibling.html

Also, we could potentially add a PerformanceTest for this pathological case with a giant DOM tree

nipunshukla21 committed:
Page freezes with many DOM manipulations and complex :has() selectors present
https://bugs.webkit.org/show_bug.cgi?id=281920
rdar://138431700

Reviewed by Matthieu Dubet and Antti Koivisto.

When evaluating :has(> :is(.x > .y)), child combinators within :is() are
scoped to the direct child's tree and don't require scope-breaking invalidation.
This avoids excessive DOM traversal on pages with many complex :has() selectors,
preventing page freezes during style recalculation.

* LayoutTests/fast/selectors/has-scope-breaking-classification-expected.txt:
* LayoutTests/fast/selectors/has-scope-breaking-classification.html:
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/invalidation/has-with-is-child-combinator-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/invalidation/has-with-is-child-combinator.html: Added.
* PerformanceTests/CSS/HasWithChildCombinatorInIs.html: Added.
* Source/WebCore/style/RuleFeature.cpp:
(WebCore::Style::computeNextHasPseudoClassMatchElement):

Canonical link: https://commits.webkit.org/302587@main
